rfd/0178-github-proxy.md

```diff
@@ -0,0 +1,497 @@
+---
+author: Steve Huang (xin.huang@goteleport.com)
+state: draft
+---
+
+# RFD 178 - GitHub Proxy
+
+## Required Approvers
+
+- Engineering: @r0mant && @smallinsky
+- Product: @klizhentas || @xinding33
+
+## What
+
+This RFD proposes design and implementation of proxying the Git SSH protocol
+for GitHub repositories.
+
+## Why
+
+GitHub Enterprise provides a security feature to bring your own SSH certificate
+authorities (CA). Once a CA is added, your organization can sign short-lived
+client SSH certificates to access organization resources on GitHub. You can
+also require your memebers to use these SSH certificates, which disables Git
+access using personal tokens.
+
+The concept of short-lived SSH certificates to access organization resources
+aligns well with Teleport, where a Teleport user begins their day with a 'tsh'
+session, accessing only what their roleset permits. Teleport can also easily
+provide the capability to issue of short-lived client SSH certificates for
+GitHub organzations so Teleport customers do not need to implement a separate
+system for issuing these certificates. 
+
+Teleport also offers other GitHub-related features, such as [GitHub IAM
+integration](https://github.com/gravitational/teleport.e/blob/master/rfd/0021e-github-iam-integration.md)
+and GitHub SSO, where this functionality can integrate nicely.
+
+## Details
+
+### UX - User stories
```

Comment:
Will the web UI also be impacted? How will a user with an enforced GitHub Proxy access model be able to interact with the GitHub web UI, for instance, to review code?

Right now, the GitHub Proxy seems to be a good fit for automation flows like CI, but for user flow missing support for GitHub web UI access will be a blocker for day-to-day work.

Reply:
RBAC should be controlled with [GitHub IAM integrartion](https://github.com/gravitational/teleport.e/blob/master/rfd/0021e-github-iam-integration.md). The proxy is more like an addon to it. As mentioned in the stories, the benefit using the GitHub Proxy include per-session MFA, audit log (e.g. generating reporting using Access Monitoring).